dbcon/joblist/tupleannexstep.cpp

```diff
@@ -512,31 +523,48 @@ void TupleAnnexStep::executeNoOrderByWithDistinct()
             for (uint64_t i = 0; i < fRowGroupIn.getRowCount() && !cancelled() && !fLimitHit; ++i)
             {
                 pair<DistinctMap_t::iterator, bool> inserted;
+                Row* rowPtr;
+
+                if (distinctMap->size() < fLimitStart)
+                    rowPtr = &rowSkip;
+                else
+                    rowPtr = &fRowOut;
+
+                if (fConstant)
+                    fConstant->fillInConstants(fRowIn, *rowPtr);
+                else
+                    copyRow(fRowIn, rowPtr);
 
-                inserted = distinctMap->insert(fRowIn.getPointer());
+                fRowIn.nextRow();
+                inserted = distinctMap->insert(rowPtr->getPointer());
                 ++fRowsProcessed;
 
                 if (inserted.second)
                 {
-                    // skip first limit-start rows
-                    if (distinctMap->size() <= fLimitStart)
-                    {
-                        fRowIn.nextRow();
-                        continue;
-                    }
-
                     if (UNLIKELY(fRowsReturned >= fLimitCount))
                     {
                         fLimitHit = true;
                         fJobList->abortOnLimit((JobStep*) this);
                         break;
                     }
 
+                    // skip first limit-start rows
+                    if (distinctMap->size() <= fLimitStart)
+                    {
+                        rowGroupSkip.incRowCount();
+                        rowSkip.nextRow();
+                        if (UNLIKELY(rowGroupSkip.getRowCount() >= rowgroup::rgCommonSize))
+                        {
+                            dataVecSkip.push_back(rgDataSkip);
+                            rgDataSkip.reinit(rowGroupSkip);
```

Comment:
We are allocating a new memory block for the RGData. Despite the fact the number of RGDatas with data to skip should be small we at least must left the commentary here.